src/Microsoft.Diagnostics.Monitoring.Egress/IEgressService.cs

```diff
@@ -0,0 +1,30 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Monitoring.Egress
+{
+    internal interface IEgressService
```

Comment:
IEgressService should be moved to RestServer.

Reply:
> I'd propose also moving EgressResult and IEgressEndpoint to the RestServer as well.

I think having egress below the rest layer makes more sense, since we'll have a triggering pipeline in the future that ideally pairs with some kind of egress. 

>Converting from pull to push is programatically simple with Stream.CopyAsync() so if every egress mechanism supported push it should be easy for the pipelines to write to it.

+1

Reply:
> IEgressService should be moved to RestServer.

Sure, I can do this for now, but we might need a concept like this when triggers are implemented, but I'll reconsider at that point.

Reply:
> I think having egress below the rest layer makes more sense, since we'll have a triggering pipeline in the future that ideally pairs with some kind of egress

My suspicion is that triggers will want some non-stream based egress. We've also got stuff like EgressResult where the Key/Value pair seems primarily intended to notify an http stream consumer that the diagnostic data was redirected somewhere else. In other words we might need an abstraction, but I think the odds are good it won't be exactly this one.

Reply:
> 1. Right now we've got two APIs to support a push model (service provides the output stream) and a pull model (caller's action provides the input stream). Can we just pick one for now rather trying to make each egress mechanism support both? Converting from pull to push is programatically simple with Stream.CopyAsync() so if every egress mechanism supported push it should be easy for the pipelines to write to it.

I've refactored all of the egress usage and implementation to use the push model and convert dump and gcdump to from pull to push as needed.

Reply:
After discussing with Wiktor, I've added the pull model back. It's now an optional implementation (overridable from the EgressEndpoint class).

This allows egress implementations to receive a stream and egress the data in more interesting ways; for example, the Azure blob storage endpoint can do parallel uploads of blocks based on the stream (faster egress) whereas the pull model only allows for sequential upload.

Another benefit is that it does not force the action callback to do stream copying (e.g. the dump collection provides a FileStream) or use intermediate streams (e.g. GCDump requires that the stream is seekable and have a known length, which isn't the case for HTTP Response streams or the stream returned by Azure blob storage BlockBlobClient.OpenWriteAsync, so an intermediate MemoryStream would be necessary to use).